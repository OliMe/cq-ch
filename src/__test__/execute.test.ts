import { execute, command } from '..';

describe('execute', () => {
  it('should create execute channel successfully', () => {
    expect(execute(['test_type'], 'test')).toBeInstanceOf(Function);
  });
  it('should create execute channel unsuccessfully', () => {
    // TODO Ignore TS errors because for raw JS we need to test argument validation.
    // @ts-ignore
    const createChannelWithoutArguments = () => execute();
    // @ts-ignore
    const createChannelWithFirstIncorrectArgument = () => execute('not array');
    const createChannelWithFirstArgumentEmptyArray = () => execute([], 'test');
    // @ts-ignore
    const createChannelWithoutSecondArgument = () => execute(['test_type']);
    // @ts-ignore
    const createChannelWithSecondIncorrectArgument = () => execute(['test_type'], 200);
    expect(createChannelWithoutArguments).toThrowErrorMatchingSnapshot();
    expect(createChannelWithFirstIncorrectArgument).toThrowErrorMatchingSnapshot();
    expect(createChannelWithFirstArgumentEmptyArray).toThrowErrorMatchingSnapshot();
    expect(createChannelWithoutSecondArgument).toThrowErrorMatchingSnapshot();
    expect(createChannelWithSecondIncorrectArgument).toThrowErrorMatchingSnapshot();
  });
});

describe('channel created by execute function', () => {
  it('should receive commands in order of sending', async () => {
    const channel = execute(['first_type', 'second_type'], 'first');
    const commandChannel = command(['first_type', 'second_type'], 'second');
    const firstTestCommand = { type: 'first_type' };
    const secondTestCommand = { type: 'second_type' };
    setTimeout(() => {
      commandChannel(secondTestCommand);
    }, 25);
    setTimeout(() => {
      commandChannel(firstTestCommand);
    }, 20);
    const firstReceivedCommand = await channel();
    expect(firstReceivedCommand).toEqual({
      ...firstTestCommand,
      context: 'second',
      timestamp: expect.any(Number),
    });
    const secondReceivedCommand = await channel();
    expect(secondReceivedCommand).toEqual({
      ...secondTestCommand,
      context: 'second',
      timestamp: expect.any(Number),
    });
  });
});
