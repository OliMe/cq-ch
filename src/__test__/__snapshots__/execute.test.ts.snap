// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute should create execute channel unsuccessfully 1`] = `"Failed to execute 'execute': first argument must be an Array with at least one element."`;

exports[`execute should create execute channel unsuccessfully 2`] = `"Failed to execute 'execute': first argument must be an Array with at least one element."`;

exports[`execute should create execute channel unsuccessfully 3`] = `"Failed to execute 'execute': first argument must be an Array with at least one element."`;

exports[`execute should create execute channel unsuccessfully 4`] = `"Failed to execute 'execute': second argument must be a string."`;

exports[`execute should create execute channel unsuccessfully 5`] = `"Failed to execute 'execute': second argument must be a string."`;
